Create JSON endpoints needed for migration source in Hugo

nick-mon1 opened the pull request:
## endpoint-scripts

This is folder contains mocks, final output json for each content types and scripts.

- `mocks/` are expected output for migration, ideal end state for the endpoint
- `endpoints/` is the actual output for each endpoint for testing/comparison
- run locally or on federalist and visit the below urls for each content type for the json endpoint 




| content      | json serializer                        | endpoint                                     | notes   |
| ------------ | -------------------------------------- | -------------------------------------------- | ------- |
| authors      | layouts/shortcodes/api-authors.html    | http://localhost:1313/authors/v1/json/       | valid   |
| news         | layouts/news/list.json.json            | http://localhost:1313/news/index.json        | valid   |
| short post   | layouts/shortcodes/api-shortposts.html | http://localhost:1313/short-posts/v1/json/   | valid   |
| events       | layouts/events/list.json.json          | http://localhost:1313/events/index.json      | valid   |
| communities  | layouts/communities/list.json.json     | http://localhost:1313/communities/index.json | valid   |
| guides       | layouts/guides/list.json.json          | http://localhost:1313/guides/index.json      | valid   |
| resources    | layouts/resources/list.json.json       | http://localhost:1313/resources/index.json   | valid   |
| topics       | layouts/topics/list.json.json          | http://localhost:1313/topics/index.json      | valid   |
| sources      | layouts/sources/list.json.json         | http://localhost:1313/sources/index.json     | valid   |
| ~~services~~ | ~~default/list.json.json~~             | ~~yes~~                                      | n/a     |
| redirects    | default/list.json.json                 |                                              | see csv |
| about        | layouts/about/list.json.json           | http://localhost:1313/about/index.json       | valid   |
| files        | layouts/files/list.json.json<br>       | http://localhost:1313/files/index.json<br>   | valid   |
| images       | layouts/images/list.json.json          | http://localhost:1313/images/index.json      | valid   |
  


  
  


## authors

| endpoint                               | serializer                           |
| -------------------------------------- | ------------------------------------ |
| http://localhost:1313/authors/v1/json/ | layouts/shortcodes/api-authors.html |


### checklist

- [x] match json fields to drupal field names in development
- [x] consistent field names in data model document
- [x] valid json?
	- [x] escaping quotes?
	- [x] returns valid
- [x] mock file complete
- [x] json response complete
- [x] capture content changes that will need to be addressed post migration 
- [ ] resolve questions on field names
	- [ ] how will `display_name` populate the title?
	- [ ] ?

### issues

- Replace single quotes to double quotes for "Abby" to 'Abby' for content/authors/abigail-abby-bowman/_index.md
- 



## communities

| endpoint                                     | serializer                         |
| -------------------------------------------- | ---------------------------------- |
| http://localhost:1313/communities/index.json | layouts/communities/list.json.json |


### checklist

- [x] match json fields to drupal field names in development
- [x] consistent field names in data model document
- [x] valid json?
	- [x] escaping quotes?
	- [x] returns valid
- [x] mock file complete
- [x] json response complete
- [x] capture content changes that will need to be addressed post migration 
- [x] resolve questions on field names


### issues

Need to support multiple platform fields:

- [ ] content/communities/artificial-intelligence.md
- [ ] content/communities/blockchain.md
- [ ] content/communities/results-oriented-accountability-for-grants.md



## events

| endpoint                                | serializer                    |
| --------------------------------------- | ----------------------------- |
| http://localhost:1313/events/index.json | layouts/events/list.json.json |


### checklist

- [x] match json fields to drupal field names in development
- [x] consistent field names in data model document
- [x] valid json?
	- [x] escaping quotes?
	- [x] returns valid
- [x] mock file complete
- [x] json response complete
- [ ] capture content changes that will need to be addressed post migration 
- [ ] resolve questions on field names
	- [ ] field_event_cop
	- [ ] venue field has sub-fields
	- [ ] youtube_title is missing



### issues





## guides

| endpoint                                | serializer                    |
| --------------------------------------- | ----------------------------- |
| http://localhost:1313/events/index.json | layouts/events/list.json.json |


### checklist

- [ ] match json fields to drupal field names in development
- [ ] consistent field names in data model document
- [ ] valid json?
	- [ ] escaping quotes?
	- [ ] returns valid
- [ ] mock file complete
- [ ] json response complete
- [ ] capture content changes that will need to be addressed post migration 
- [ ] resolve questions on field names
	- [ ] 


### issues

- [ ] not reading guide root index files for some reason
- [ ] add `field_guide_nav_title` to front matter for each guide
- [ ] landing page?
- [ ] navigation page?




## news

| endpoint                              | serializer                  |
| ------------------------------------- | --------------------------- |
| http://localhost:1313/news/index.json | layouts/news/list.json.json |


### checklist

- [x] match json fields to drupal field names in development
- [x] consistent field names in data model document
- [x] valid json?
	- [x] escaping quotes?
	- [x] returns valid
- [x] mock file complete
- [x] json response complete
- [x] capture content changes that will need to be addressed post migration 
- [x] resolve questions on field names


### issues


## resources

| endpoint                                   | serializer                       |
| ------------------------------------------ | -------------------------------- |
| http://localhost:1313/resources/index.json | layouts/resources/list.json.json |


### checklist

- [x] match json fields to drupal field names in development
- [x] consistent field names in data model document
- [x] valid json?
	- [x] escaping quotes?
	- [x] returns valid
- [x] mock file complete
- [x] json response complete
- [x] capture content changes that will need to be addressed post migration 
- [x] resolve questions on field names


### issues

- [x] missing field_authors in drupal


## sources

| endpoint                                 | serializer                     |
| ---------------------------------------- | ------------------------------ |
| http://localhost:1313/sources/index.json | layouts/sources/list.json.json |


### checklist

- [x] match json fields to drupal field names in development
- [x] consistent field names in data model document
- [x] valid json?
	- [x] escaping quotes?
	- [x] returns valid
- [x] mock file complete
- [x] json response complete
- [x] capture content changes that will need to be addressed post migration 
- [ ] resolve questions on field names


### issues

- [ ] missing drupal field names

## topics type

| endpoint                             | serializer                    |
| ------------------------------------ | ----------------------------- |
| http://localhost:1313/topics/v1/json | layouts/events/list.json.json |


### checklist

- [x] match json fields to drupal field names in development
- [x] consistent field names in data model document
- [x] valid json?
	- [x] escaping quotes?
	- [x] returns valid
- [x] mock file complete
- [x] json response complete
- [x] capture content changes that will need to be addressed post migration 
- [ ] resolve questions on field names


### issues

- how will we ingest featured communities, featured links, featured resources, and legislation

mattsqd committed:
Added a couple helper scripts to start the site and run command to start the site without livereload so that JSON files could be parsed.

mattsqd committed:
Prepare image JSON shortcode to be migratable to Drupal. Added new source-url field, dont load image if it doesnt have a uid or format, and reformat to read more easily.

ToniBonittoGSA committed:
Merge branch 'main' into 1631-migrate-approaches

ToniBonittoGSA committed:
Merge branch 'main' into 1631-migrate-approaches

mattsqd committed:
Created a sample page that lists out an example of all shortcodes.

nick-mon1 committed:
First steps for json endpoints

- added missing fields
- removed duplicate keys
- added shortposts endpoint

nick-mon1 committed:
Reviewed all json endpoints

mattsqd committed:
Merge remote-tracking branch 'origin/main' into shortcode-examples

ToniBonittoGSA committed:
Merge branch 'main' into shortcode-examples

nick-mon1 committed:
Refactored endpoints, authors and events are done

nick-mon1 committed:
Consolidated to README.md

nick-mon1 committed:
5 content types setup

nick-mon1 committed:
Merge branch 'main' into nl-json-endpoints

nick-mon1 committed:
Removed sass and sharp arm64 versions

nick-mon1 committed:
Merge branch 'nl-json-endpoints' of github.com:GSA/digitalgov.gov into nl-json-endpoints

nick-mon1 committed:
Removed sass

nick-mon1 committed:
Fixed loop logic for news, resources

nick-mon1 committed:
Fixed topics, communities endpoints

nick-mon1 committed:
Fixed formatting updates

nick-mon1 committed:
Fixed sass package platform

nick-mon1 committed:
Added topics front matter fields, fixed links

nick-mon1 committed:
Updated readme with matt notes

mattsqd committed:
Merge remote-tracking branch 'origin/main' into shortcode-examples

nick-mon1 committed:
Updated readme.md notes for redirects work

mattsqd committed:
Change aliases to redirects, since that is what they are in Drupal.

mattsqd committed:
Don't worry about empty fields for unique fields.

mattsqd committed:
Add in '.Type' to all references since some reference fields might hold multiple types.

mattsqd committed:
Create a uid that is type + slug for all endpoints.

mattsqd committed:
Add on .Type to topics references.

mattsqd committed:
Undo adding .Type onto references and uid.

mattsqd committed:
Undo temp changes to an author.

mattsqd committed:
Merge remote-tracking branch 'origin/1631-migrate-approaches' into nl-json-endpoints

# Conflicts:
#	package.json

mattsqd commented:
@nick-mon1 I merged my branch (https://github.com/GSA/digitalgov.gov/pull/7840/files) that updates the endpoint for images into this one so we have all the migration stuff in one branch.
I also:
* added uid field to all endpoints that is slug and if that's empty, defaults to .Name if slug is empty, just in case.
* Used relative link instead of permanent link for every endpoints URL field.
* Changed aliases endpoint to redirects, since that is how they are treated in Drupal.

This is looking great, I can't wait to start testing this against actually Drupal migrations. I'm not going to merge this and instead just leave this as a long living branch as we update it with subsequent tasks.

nick-mon1 committed:
Added endpoint for guides

nick-mon1 committed:
Got guides in proper format, need to check some fields, flag files for data cleanup

nick-mon1 committed:
Finalized content type endpoints for testing

nick-mon1 committed:
Merge branch 'nl-json-endpoints' of github.com:GSA/digitalgov.gov into nl-json-endpoints

nick-mon1 committed:
Merge branch 'main' into nl-json-endpoints

nick-mon1 committed:
Fixed json formatting for topics

nick-mon1 committed:
Merge branch 'nl-json-endpoints' of github.com:GSA/digitalgov.gov into nl-json-endpoints

nick-mon1 committed:
Updated short-posts endpoint, guides and content cleanup

nick-mon1 committed:
Minor template format fixes

mattsqd committed:
Removed shortcodes not being migrated. Changed YouTube ref to one created in Drupal already so a new one doesn't get created.

mattsqd committed:
Merge remote-tracking branch 'origin/main' into shortcode-examples

mattsqd committed:
Merge branch 'shortcode-examples' of github.com:GSA/digitalgov.gov into shortcode-examples

nick-mon1 committed:
Removed additional tweet removals

mattsqd commented:
I'm getting a JSON parse error on [topics](http://localhost:1313/topics/v1/json/).
I think if we `| jsonify` all the values it won't happen. For example, it's [field_summary ](https://github.com/GSA/digitalgov.gov/blob/743ede775543576d50886bb39b5100008fbdd966/themes/digital.gov/layouts/partials/api/metadata.html#L11)that is causing topics to not be valid

mattsqd commented:
Where is the CSV that you mentioned for the redirects? I also am not sure where the code is you used to generate it (I might be blind).
`default/list.json.json` Seems to just return all top level pages. Do we have an end point for 'Basic Pages' yet? For us, that would be everything in the `content/about` directory. The other thing we'll need, is an endpoint for `data/files`, which should be pretty easy since it's nearly identical to the images endpoint.

mattsqd reviewed (changes requested):
The endpoints are coming alive! I just had a few points and questions in some comments.

mattsqd commented:
For the authors endpoint:
* abigail-abby-bowman does not have a field_display_name
* For some reason, the author landing page is the first item in the endpoint

nick-mon1 committed:
Replaced legacy images/files, moved assets to root on s3

nick-mon1 committed:
Removed broken tweets

nick-mon1 committed:
Merge branch 'main' into nl-json-endpoints

nick-mon1 committed:
Fixed topics and redirects formatting

nick-mon1 committed:
Merge branch 'nl-json-endpoints' of github.com:GSA/digitalgov.gov into nl-json-endpoints

mattsqd committed:
Add 'from' sub key for redirects because of a quirk in how Drupal migrates.

mattsqd committed:
Added back in trailing new line.

nick-mon1 committed:
Updated endpoints and moved dupe field for featured resources

- [x] added /files/index.json
- [x] added /images/index.json
- [x] added check for _index.md files for /guides/index.json

nick-mon1 committed:
Made featured_resources an array

nick-mon1 commented on themes/digital.gov/layouts/partials/api/topics-metadata.html:
```diff
@@ -0,0 +1,70 @@
+{{- end -}}
+{{- if (isset .Params "featured_resources") -}}
+  {{- range .Params.featured_resources.resources -}}
+    "field_featured_resource": "{{ .link }}",
```
See https://github.com/GSA/digitalgov.gov/blob/4c5e3df9aabddbfdda976fa0dc421149daeda0ff/endpoint-scripts/mocks/topics-mock.json.

mattsqd committed:
Merge remote-tracking branch 'origin/shortcode-examples' into nl-json-endpoints

# Conflicts:
#	content/news/2020/06/2020-06-30-bringing-our-humanity-work-5-ingredients.md

mattsqd committed:
Update shortcodes to match the new types added to drupal and deprecated ones removed.

mattsqd committed:
Metadata is usually 0, so must use isset.

mattsqd committed:
Field venue needed to be an object. Added event platform.

nick-mon1 committed:
Added .Content to resources for testing

nick-mon1 committed:
Merge branch 'nl-json-endpoints' of github.com:GSA/digitalgov.gov into nl-json-endpoints

nick-mon1 committed:
Setup field_body for endpoints

nick-mon1 commented:
@mattsqd I've added the `field_body` to all content types. It works for all content types except news, there is a break somewhere possibly related to markdown formatting, shortcodes parsing or an improperly formatted file.

I noticed that for events ([link](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.sites.pages.cloud.gov/preview/gsa/digitalgov.gov/nl-json-endpoints/events/index.json)) there are newlines within html tags. I think this is related to the markdown parsing some shortcodes.

Good news is that once we merge https://github.com/GSA/digitalgov.gov/pull/8320 it should make it easier to address issues. We are shooting for mid to late next week to get that ready.

nick-mon1 committed:
Minor updates

nick-mon1 committed:
Renamed communities filenames to match slug for consistency

nick-mon1 committed:
Updated logo path for source endpoint

mattsqd commented:
@nick-mon1 we have a topic and a community that are trying to be the source of the same redirect
![image](https://github.com/user-attachments/assets/1430f470-f6b4-4a22-b8ff-c6cdbf6189aa)
Can we remove the alias from one, please?

nick-mon1 committed:
Updated slugs to match filenames

nick-mon1 committed:
updated featured_resources to match featured_links for internal references

nick-mon1 committed:
Fixed html table formatting for 2013-11-01-federal-government-challenges-and-contests

nick-mon1 committed:
Removed federalist path from links

nick-mon1 committed:
Fix #1 for removing base path

nick-mon1 committed:
Fixed replace logic

nick-mon1 committed:
Fixed replace logic

nick-mon1 committed:
Merge branch 'nl-json-endpoints' of github.com:GSA/digitalgov.gov into nl-json-endpoints

nick-mon1 committed:
Removed /content/guides/_index.md

nick-mon1 committed:
Fixed broken link

nick-mon1 committed:
Merge branch 'main' into nl-json-endpoints

nick-mon1 committed:
Removed markdown rendering

nick-mon1 committed:
Fix markdown expirydate formattig

nick-mon1 committed:
Fixed frontmatter expirydate tags for news posts

nick-mon1 committed:
Merge branch 'nl-fix-news-expirydate-formatting' into nl-json-endpoints

nick-mon1 committed:
Removed spaces between markdown link text and link url

nick-mon1 committed:
Fixed quote escape typo

nick-mon1 committed:
Removed unused yaml files

nick-mon1 committed:
Removed broken links from archival

nick-mon1 committed:
Added yml files for missing files for endpoints

nick-mon1 committed:
Moved legacy images to img shortcode format

nick-mon1 committed:
Fixed duplicate aliases, added missing yml for files endpoint

nick-mon1 committed:
Removed external redirected pages for news, resources

nick-mon1 committed:
Updated featured resources and added missing yml file

nick-mon1 committed:
Updated file names to be consistent

nick-mon1 committed:
Removed double alias and self referencing alias

mattsqd committed:
Added more examples of ref and rel.

mattsqd committed:
Merge branch 'nl-json-endpoints' of github.com:GSA/digitalgov.gov into nl-json-endpoints

nick-mon1 committed:
Updated slugs for authors and opm-labs source
